### PR TITLE
Initial Setup

### PR DESCRIPTION
**Pull request by elcritch**
This makes an async compatible queue, see `queues.nim`. Then `jobs.nim` builds on this with a type and macro for submitting jobs to taskpools and getting a result back via an async future. Finally `tasks.nim` automates the jobs wrapping and adds some type extra type checking.

Here's an example of the transformations for an "asyncTasklet":

```nim
proc addNums(a, b: float): float {.asyncTask.} =
  os.sleep(100)
  echo "adding: ", a, " + ", b
  return a + b
```
becomes:
```nim
proc addNumsTasklet(a, b: float): float {.nimcall.} =
  os.sleep(100)
  echo "adding: ", a, " + ", b
  return a + b

proc addNums(jobResult: JobResult[float]; a, b: float) {.nimcall.} =
  let val {.inject.} = addNumsTasklet(checkParamType(a), checkParamType(b))
  discard jobResult.queue.send((jobResult.id, val))

```

Calling `addNums`:
```nim
    let res = await jobs.submit(addNums(1.0, 2.0,))
```
becomes:
```nim
let (jobRes_503318661, fut_503318662) = createFuture(jobs, "addNums(1.0, 2.0)")
jobs.taskpool.spawn(addNums(jobRes_503318661, 1.0, 2.0))
fut_503318662
```

### CODE REVIEW

**Comment by elcritch** [comment]
Hey @tbekas finally got a PR up. It's fairly complete but there's a few corner cases to figure out. 

The openArray seem to work, but will need to verify the GC won't free the main GC data.

**Comment by elcritch** [comment]
This will need more tests, stress testing to verify the threading. It'd be nice to do some basic benchmarking as well. 

Also the API's likely need some changes. In particular the jobs `newJobQueue` should probably return a result type.

File: apatheia.nimble [hunk 1]
```diff
@@ -6,7 +6,10 @@ description   = "Async support for threading primitives"
 license       = "MIT"
 srcDir        = "src"
 
-
 # Dependencies
 
-requires "nim >= 1.6.18"
+requires "chronos >= 4.0.0"
+requires "threading"
+requires "taskpool"
```
Comment (tbekas) [inline apatheia.nimble]: Nimble fails to resolve this dependency, is it used? If I comment it out all tests are passing.

File: src/apatheia/jobs.nim [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  ## Starts a "detached" async processor for a given job queue.
+  ## 
+  ## This processor waits for events from the queue in the JobQueue
+  ## and complete the associated futures.
+
+  while jobs.running:
+    echo "jobs running..."
+    let res = await(jobs.queue.wait()).get()
+    echo "jobs result: ", res.repr
+    echo "jobs futes: ", jobs.futures.unsafeAddr.pointer.repr, " => ", jobs.futures.keys().toSeq()
```
Comment (tbekas) [inline src/apatheia/jobs.nim]: Maybe add a chronicles dependency and use trace/debug for this?

**Review by tbekas — COMMENTED** [review]
Overall a great job 👍 This PR would use some cleanup of the commented and unused code as well as some echos.

File: src/apatheia/jobs.nim (hunk 2 above)
Comment (elcritch) [reply src/apatheia/jobs.nim]: @tbekas that's a good idea. Some of this is just initial debugging and to be removed, but we'll definitely want to log certain events.

File: apatheia.nimble (hunk 1 above)
Comment (elcritch) [reply apatheia.nimble]: Thanks, I'm using nim-taskpool, but I may have gotten the name wrong. Ok, I'll check it out.

File: tests/ttasks.nim [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    echo "\nstart"
+    let res = await jobs.submit(addNums(1.0, 2.0,))
+    echo "result: ", res.repr
+    check true
+
+  asyncTest "test addNumValues":
+    var jobs = newJobQueue[float](taskpool = tp)
+    echo "\nstart"
+    let args = @[1.0, 2.0, 3.0]
+    let res = await jobs.submit(addNumValues(args))
```
Comment (elcritch) [inline tests/ttasks.nim:32]: @tbekas this api currently doesn't support back-pressure. However making `submit` async adds overhead and just changes the queue filling to Chronos's queue.

Instead, what do you think about an additional api `trySubmit(...): Option[Future[T]]` which would only submit and return the future if the queue isn't full. Then the upstream code could handle / reject processing jobs appropriately.

The other alternative would be throwing an exception when the queue is full, though that wouldn't fit the Codex codebase as well.

Comment (elcritch) [reply tests/ttasks.nim:32]: @tbekas I'm going to merge this PR. Another can be opened for this.

Comment (tbekas) [reply tests/ttasks.nim:32]: I like the `trySubmit(...): Option[Future[T]]` API 👌
